common/elasticsearch/validator/searchAttrValidator.go

```diff
@@ -73,12 +74,19 @@ func (sv *SearchAttributesValidator) ValidateSearchAttributes(input *gen.SearchA
 	}
 
 	totalSize := 0
+	validAttr := sv.validSearchAttributes()
 	for key, val := range fields {
 		// verify: key is whitelisted
-		if !sv.isValidSearchAttributes(key) {
+		if !sv.isValidSearchAttributesKey(validAttr, key) {
+			sv.logger.WithTags(tag.ESKey(key), tag.WorkflowDomainName(domain)).
+				Error("invalid search attribute key")
+			return &gen.BadRequestError{Message: fmt.Sprintf("%s is not a valid search attribute key", key)}
+		}
+		// verify: value has the correct type
+		if !sv.isValidSearchAttributesValue(validAttr, key, val) {
 			sv.logger.WithTags(tag.ESKey(key), tag.WorkflowDomainName(domain)).
-				Error("invalid search attribute")
-			return &gen.BadRequestError{Message: fmt.Sprintf("%s is not valid search attribute", key)}
+				Error(fmt.Sprintf("%s is not a valid search attribute value for key %s", val, key))
```

Comment:
let's use tags instead of formatted message.

Reply:
Done.